samples/sample_common/include/sample_utils.h

```diff
@@ -495,6 +495,53 @@ mfxF64 CalculateFrameRate(mfxU32 nFrameRateExtN, mfxU32 nFrameRateExtD);
 mfxU16 GetFreeSurfaceIndex(mfxFrameSurface1* pSurfacesPool, mfxU16 nPoolSize);
 mfxU16 GetFreeSurface(mfxFrameSurface1* pSurfacesPool, mfxU16 nPoolSize);
 void FreeSurfacePool(mfxFrameSurface1* pSurfacesPool, mfxU16 nPoolSize);
+
+class mfxBitstreamWrapper : public mfxBitstream
+{
+public:
+    mfxBitstreamWrapper()
+        : mfxBitstream()
+    {}
+
+    mfxBitstreamWrapper(mfxU32 n_bytes)
+        : mfxBitstream()
+    {
+        Extend(n_bytes);
+    }
+
+    mfxBitstreamWrapper(const mfxBitstreamWrapper & bs_wrapper)
+    {
+        m_data = bs_wrapper.m_data;
+
+        static_cast<mfxBitstream&>(*this) = static_cast<const mfxBitstream&>(bs_wrapper);
+
+        Data = m_data.data();
+    }
+
+    mfxBitstreamWrapper & operator= (mfxBitstreamWrapper bs_wrapper)
+    {
+        std::swap(*this, bs_wrapper);
```

Comment:
`swap` will use `operator=` isn't it?
Looks like we need either to define this operator as `operator= (mfxBitstreamWrapper const& bs_wrapper)` and have `operator= (mfxBitstreamWrapper&& bs_wrapper) = default` or swap class' members one by one (like `swap(this->m_data,  bs_wrapper.m_data)` or define our own `swap` for _mfxBitstreamWrapper_ class

Reply:
I think it should use move assignment operator... will check it out

Reply:
@mgonchar,

Yes, it should. But u define your own copy assignment operator - AFAIR this prevents implicitly declared move assignment being generated.

Reply:
yes, you are right. Right solution here would be to implement copy-swap idiom via several swaps like:
```
mfxBitstreamWrapper & operator= (mfxBitstreamWrapper bs_wrapper)
{
    std::swap(static_cast<mfxBitstream&>(*this), static_cast<const mfxBitstream&>(bs_wrapper));
    std::swap(m_data, bs_wrapper.m_data);
}
```

But as for me better to default move assignment and implement fair copy assignment here

Reply:
> But as for me better to default move assignment and implement fair copy assignment here

copy-swap gives two benefits - exception safety (since swap is noexcept) and protection against self-assignment. So, I'd rather use copy-swap.

Reply:
Well, exception safety of copy-swap is exactly same as for fair copy assignment (both of them can throw ```std::bad_alloc``` at allocation of new chunk of memory). But for copy-swap we lose defaulted move assignment, and we have temporary object for move assignment, which we don't have in case of default move assignment. Now what related to copy assignment, as I said we remain the same level of exception safety in this particular case and also we won't have temporary object. My point to introduce copy-swap was to reduce code duplication and for ```std::unique_ptr``` design, which I used at first time, it was critical, because it involved low-level memory copying, which I didn't want to duplicate) But with ```std::vector``` and such complicate (same amount of code lines as fair assignment) copy-swap assignment operator I don't see any benefits in it for this particular case. Also I like this idiom, but as for me it doesn't fit good this particular case

Reply:
> exception safety of copy-swap is exactly same as for fair copy assignment

In my understanding, no, it's not the same. For copy-swap your object either successfully copied or u get an exception and object remain at the same state and still could be used. For this "fair copy assignment" your object might be at the intermediate sate - `mfxBitstream` subobject was updated and points to the memory that actually belongs to the another object. The first invariant is - `Data` points to the memory that is owned by object's vector (`m_data` member). When we have `bad_alloc` during `m_data = bs_wrapper.m_data` this variant is violated and object can't be used anymore.

I'd rather do something like
```
mfxBitstreamWrapper& operator=(mfxBitstreamWrapper  const& other)
{
    mfxBitstreamWrapper tmp(other);
    using std::swap;
    swap(tmp.m_data, this->m_data);
    swap(static_cast<mfxBitstream&>(tmp), static_cast<mfxBitstream&>(*this));

    return *this;
}

mfxBitstreamWrapper & operator= (mfxBitstreamWrapper&&) = default; 
```

Reply:
Agree, we can't get more than basic exception safety guarantee here without intermediate copy. I also think we can save some tmp copies (which arises inside ```std::swap``` like this):
```
mfxBitstreamWrapper& operator=(mfxBitstreamWrapper  const& other)
{
    mfxBitstreamWrapper tmp(other);

    m_data = std::move(tmp.m_data);
    static_cast<mfxBitstream&>(*this) = static_cast<mfxBitstream&>(tmp);

    return *this;
}

mfxBitstreamWrapper & operator= (mfxBitstreamWrapper&&) = default; 
```

Reply:
For me, `swap` is more idiomatic. Anyway, current variant is quite OK, so let's move forward.